src/lightning/fabric/strategies/model_parallel.py

```diff
@@ -310,3 +358,206 @@ def __enter__(self) -> None:
+
+        requested_metadata_keys = state.keys() - modules.keys() - optimizers.keys()
+        _validate_keys_for_strict_loading(requested_metadata_keys, checkpoint.keys(), strict=strict)
+
+        # Load metadata (anything not a module or optimizer)
+        _move_state_into(source=checkpoint, destination=state, keys=requested_metadata_keys)
+
+        # return the remaining metadata that wasn't requested as part of `state`
+        return checkpoint
+
+    raise ValueError(
+        f"The path {str(path)!r} does not point to a valid checkpoint. Make sure the path points to either a"
+        " directory with distributed checkpoint shards, or a single file with a full checkpoint."
+    )
+
+
+def _has_dtensor_modules(module: object) -> TypeGuard[Module]:
+    from torch.distributed._tensor import DTensor
+
+    return isinstance(module, Module) and any(isinstance(t, DTensor) for t in module.parameters())
+
+
+def _load_raw_module_state_from_path(path: Path, module: Module, world_size: int, strict: bool = True) -> None:
+    """Loads the state dict from a file path into the FSDP module."""
+    if not _is_full_checkpoint(path):
+        raise ValueError(
+            "Failed to load checkpoint directly into the model. The given path must be a single file containing the"
+            f" full state dict: {path}"
+        )
+    # Use `lazy_load` instead of `torch.load` here to avoid storing a copy of the full checkpoint per rank
+    _load_raw_module_state(state_dict=_lazy_load(path), module=module, world_size=world_size, strict=strict)
+
+
+def _load_raw_module_state(state_dict: Dict[str, Any], module: Module, world_size: int, strict: bool = True) -> None:
+    """Loads the state dict into the module by gathering all weights first and then and writing back to each shard."""
+    from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+
+    if _TORCH_GREATER_EQUAL_2_3 and _has_dtensor_modules(module):
+        _load_dtensor_state_dict(state_dict, module=module, strict=strict)
+    elif isinstance(module, FSDP):
+        with _get_full_state_dict_context(module, world_size=world_size, rank0_only=False):
+            module.load_state_dict(state_dict, strict=strict)
+    else:
+        module.load_state_dict(state_dict, strict=strict)
+
+
+def _load_dtensor_state_dict(state_dict: Dict[str, Any], module: Module, strict: bool = True) -> None:
+    # Note: There is currently no standard way in PyTorch to load a single-file checkpoint into a
+    # DTensor module and/or optimizer. We manually reshard and assign the tensors to the model, but don't
+    # support loading the optimizer state yet. Caveat: Custom `Module.load_state_dict()` methods are not supported
+    from torch.distributed._tensor import DTensor, distribute_tensor
+
+    module_keys = {name for name, _ in itertools.chain(module.named_parameters(), module.named_buffers())}
```

Comment:
I think named_buffers includes non-persistent buffers. This would break under strict-loading. You'll have to filter them out from this